README.md
# TableFlip

### (╯°□°）╯︵ ┻━┻

### ┬──┬ ノ( ゜-゜ノ)
---

[![Pod Version](https://img.shields.io/badge/Pod-1.1-6193DF.svg)](https://cocoapods.org/)
![Swift Version](https://img.shields.io/badge/Swift-4.2-brightgreen.svg)
![License MIT](https://img.shields.io/badge/License-MIT-lightgrey.svg) 
![Plaform](https://img.shields.io/badge/Platform-iOS-lightgrey.svg)

### Animations are cool. UITableView isn't. So why not make animating UITableView cool?

---

##### The entire API for TableFlip consists of an animation, and a call to animate. It's as simple as that.

##### TableFlip takes the annoying nature of reloading UITableViews, maintaining state, and anminating, and minimizes it to two lines of code.

##### If you want to animate every cell consecutively, the code will look like this.

```swift
self.tableView.reloadData()
self.tableView.animate(animation: myCoolCellAnimation)
```

##### And if you want to animate the entire table view at once, the code will look like this: 

```swift
self.tableView.reloadData()
self.tableView.animate(animation: myCoolTableAnimation)
```

Animations are completely customizable and configurable. TableFlip provides a few built-in defaults that work nicely out the box, if you don't want to make your own.

## Let's look at some examples


### Animating from the left with a nice staggered effect:

```swift
TableViewAnimation.Cell.left(duration: 0.5)
```

![](gifs/left.gif)


### Pushing from the top can make a nice initial loading effect:

```swift
TableViewAnimation.Table.top(duration: 0.8)
```

![](gifs/top.gif)

### A simple fade is always elegant:

```swift
TableViewAnimation.Cell.fade(duration: 1.0)
```

![](gifs/fade.gif)


### And you can make your own transform, as fun or weird as you want by using CGAffineTransform:

```swift
let degrees = sin(90.0 * CGFloat.pi/180.0)
let rotationTransform = CGAffineTransform(rotationAngle: degrees)
let flipTransform = CGAffineTransform(scaleX: -1, y: -1)
let customTransform = rotationTransform.concatenating(flipTransform)

let customAnimation = TableViewAnimation.Cell.custom(duration: 0.6, transform: customTransform, options: .curveEaseInOut)
                
self.tableView.animate(animation: customAnimation, completion: nil)
```

![](gifs/custom.gif)

## Requirements

- iOS 8.0+
- Xcode 10.0+
- Swift 4.2

## Installation
You can use [CocoaPods](http://cocoapods.org/) to install `TableFlip` by adding it to your `Podfile`:

```ruby
platform :ios, '8.0'
use_frameworks!

pod 'TableFlip'
```

Or install it manually by downloading `UITableView+Animations.swift` and dropping it in your project.

## About me

Hi, I'm [Joe](http://fabisevi.ch) everywhere on the web, but especially on [Twitter](https://twitter.com/mergesort).

## License

See the [license](LICENSE) for more information about how you can use TableFlip.

## P.S.

If all you liked was the kitty, that's my baby Gif. Feel free to follow her on the [Instagram](https://www.instagram.com/gifthecat).


CHANGELOG.md
# Change Log

# 1.3 (2018-12-31)

- Added support for Swift 4.2.

# 1.2 (2017-09-27)

- `animateTableView()` and `animateCells()` have been renamed to `animate()` and are differentiated based on the parameter passed in.

- Adding a completion handler when `reloadData(inPlace: animated:)` is called synchronously.

- `reloadData(smoothly: animated:)` has been renamed `reloadData(inPlace: animated:)`

- A fix for `animateWithFade()` potentially calling it's completion block multiple times.

# 1.1 (2017-08-27)

This release updates behavior that cell animations will fire their completion block after all the cell animations have completed, not after each individual animation.

It is not a minor semantic version bump because while it is a bug fix, it produces a change in behavior which some people may have been relying on.


# 1.0.1 (2017-08-27)

This update adds an `indexPaths` parameter to `func animateCells`. 

~`func animateCells(animation: TableViewAnimation.Cell, completion: (() -> Void)? = nil)`~

`func animateCells(animation: TableViewAnimation.Cell, indexPaths: [IndexPath]? = nil, completion: (() -> Void)? = nil)`

This change is non-breaking, since `indexPaths` parameter is optional. Your old code should continue to work as it did before.



# 1.0 (2016-11-15)

Initial release.


Example/TableFlipExample/TableViewController.swift
//
//  TableViewController.swift
//  TableFlipExample
//
//  Created by Joe Fabisevich on 8/27/17.
//  Copyright © 2017 Mergesort. All rights reserved.
//

import UIKit

enum ExampleAnimation {

    case fade
    case top
    case left
    case custom
    case indexPaths

}

final class TableViewController: UIViewController {

    private let animation: ExampleAnimation

    private let dataSource = ExampleDataSource()
    
    let tableView: UITableView = {
        let tableView = UITableView()
        tableView.rowHeight = 64.0
        tableView.tableFooterView = UIView()
        
        return tableView
    }()


    // MARK: Initializers

    init(animation: ExampleAnimation) {
        self.animation = animation

        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("Unimplemented, and loving it.")
    }


    // MARK: Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.setup()
    }
}

private extension TableViewController {
    
    func setup() {
        self.tableView.dataSource = self.dataSource

        self.view.addSubview(self.tableView)

        self.setupConstraints()

        switch self.animation {

        case .custom:
            self.title = "Custom Animation"
            
        case .top:
            self.title = "Top Animation"
            
        case .left:
            self.title = "Left Animation"
            
        case .fade:
            self.title = "Fade Animation"
            
        case .indexPaths:
            self.title = "Select Index Paths Animation"
            
        }
        
        // Delayed to simulate a network connection, like in the real world, which is where I live.
        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
            self.dataSource.hasLoaded = true
            self.tableView.reloadData()
            self.animateTableView()
        }
    }
    
    func setupConstraints() {
        self.tableView.translatesAutoresizingMaskIntoConstraints = false

        NSLayoutConstraint.activate([
            self.tableView.leadingAnchor.constraint(equalTo: self.view.leadingAnchor),
            self.tableView.trailingAnchor.constraint(equalTo: self.view.trailingAnchor),
            self.tableView.topAnchor.constraint(equalTo: self.view.topAnchor),
            self.tableView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor),
        ])
    }
    
    func animateTableView() {
        switch self.animation {

        case .top:
            let topAnimation = TableViewAnimation.Table.top(duration: 0.8)
            self.tableView.animate(animation: topAnimation)

        case .fade:
            let fadeAnimation = TableViewAnimation.Cell.fade(duration: 0.5)
            self.tableView.animate(animation: fadeAnimation)

        case .left:
            let leftAnimation = TableViewAnimation.Cell.left(duration: 0.5)
            self.tableView.animate(animation: leftAnimation, indexPaths: nil, completion: nil)

        case .custom:
            let degrees = sin(90.0 * CGFloat.pi/180.0)
            let rotationTransform = CGAffineTransform(rotationAngle: degrees)
            let flipTransform = CGAffineTransform(scaleX: -1, y: -1)
            let customTransform = rotationTransform.concatenating(flipTransform)
            let customAnimation = TableViewAnimation.Cell.custom(duration: 0.6, transform: customTransform, options: .curveEaseInOut)

            self.tableView.animate(animation: customAnimation, completion: nil)

        case .indexPaths:
            let groupedItems = Dictionary(grouping: (0..<self.dataSource.exampleItems.count), by: { $0 % 2 })

            let oddIndices = groupedItems[1]?.compactMap { IndexPath(row: $0, section: 0) }
            let leftAnimation = TableViewAnimation.Cell.left(duration: 0.5)
            self.tableView.animate(animation: leftAnimation, indexPaths: oddIndices, completion: nil)

            let evenIndices = groupedItems[0]?.compactMap { IndexPath(row: $0, section: 0) }
            let rightAnimation = TableViewAnimation.Cell.right(duration: 0.5)
            self.tableView.animate(animation: rightAnimation, indexPaths: evenIndices, completion: nil)
        }
    }
    
}

private struct CellData {
    
    let title: String
    let subtitle: String
    let image: UIImage?

}

private final class ExampleDataSource: NSObject, UITableViewDataSource {

    var hasLoaded: Bool = false

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return hasLoaded ? self.exampleItems.count : 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell(style: .subtitle, reuseIdentifier: nil)

        let currentCellData = self.exampleItems[indexPath.row]

        cell.textLabel?.text = currentCellData.title
        cell.detailTextLabel?.text = currentCellData.subtitle
        cell.imageView?.image = currentCellData.image

        return cell
    }

    let exampleItems: [CellData] = [
        CellData(title: "Playing Astronaut", subtitle: "Age: 12 months", image: #imageLiteral(resourceName: "playing_astronaut")),
        CellData(title: "Fetching", subtitle: "Age: 3 months", image: #imageLiteral(resourceName: "fetching")),
        CellData(title: "Being Adorable", subtitle: "Age: 7 months", image: #imageLiteral(resourceName: "being_adorable")),
        CellData(title: "Buying Sneakers", subtitle: "Age: 10 months", image: #imageLiteral(resourceName: "buying_sneakers")),
        CellData(title: "Napping", subtitle: "Age: 4 months", image: #imageLiteral(resourceName: "napping")),
        CellData(title: "Breathing Deeply", subtitle: "Age: 9 months", image: #imageLiteral(resourceName: "breathing_deeply")),
        CellData(title: "Camouflaging", subtitle: "Age: 8 months", image: #imageLiteral(resourceName: "camouflaging")),
        CellData(title: "Screaming Her Butt Off", subtitle: "Age: 2 months", image: #imageLiteral(resourceName: "screaming_her_but_off")),
    ]

}
